File: scripts/fix_front_matter.py
import os
import yaml
import re
from pathlib import Path

DOCS_DIR = Path('docs')
SCHEMA_PATH = Path('schemas/front_matter.schema.json')

KIND_MAP = {
    'explanation': 'explanation',
    'how-to': 'how-to',
    'reference': 'reference',
    'tutorial': 'tutorial',
    'design': 'design',
    'concept': 'concept',
    'requirements': 'requirements',
    'adr': 'adr'
}

AREA_MAP = {
    'flight-core': 'flight-core',
    'flight-virtual': 'flight-virtual',
    'flight-hid': 'flight-hid',
    'flight-ipc': 'flight-ipc',
    'flight-scheduler': 'flight-scheduler',
    'flight-ffb': 'flight-ffb',
    'flight-panels': 'flight-panels',
    'infra': 'infra',
    'infrastructure': 'infra',
    'ci': 'ci',
    'simulation': 'simulation',
    'integration': 'integration',
    'ksp': 'ksp',
    'profile': 'profile'
}

def get_default_kind(path):
    parts = path.parts
    if 'explanation' in parts:
        if 'adr' in parts:
            return 'adr'
        return 'explanation'
    if 'how-to' in parts:
        return 'how-to'
    if 'reference' in parts:
        return 'reference'
    if 'tutorials' in parts:
        return 'tutorial'
    if 'design' in parts:
        return 'design'
    if 'dev' in parts:
        return 'explanation'
    return 'concept'

def get_default_area(path, content):
    # Try to infer area from content or path
    content_lower = content.lower()
    for area in AREA_MAP.values():
        if area in content_lower:
            return area
    return 'flight-core' # Default

def slugify(name):
    return re.sub(r'[^A-Z0-9]+', '-', name.upper()).strip('-')

def fix_file(path):
    with open(path, 'r', encoding='utf-8') as f:
        content = f.read()

    front_matter = {}
    body = content
    
    match = re.match(r'^---\s*\n(.*?)\n---\s*\n', content, re.DOTALL)
    if match:
        try:
            front_matter = yaml.safe_load(match.group(1))
            body = content[match.end():]
        except yaml.YAMLError:
            print(f"Error parsing YAML in {path}")
            return

    # Normalize fields
    if 'category' in front_matter:
        front_matter['kind'] = KIND_MAP.get(front_matter.pop('category'), 'concept')
    
    if 'group' in front_matter:
        front_matter['area'] = AREA_MAP.get(front_matter.pop('group'), 'flight-core')

    if 'kind' not in front_matter:
        front_matter['kind'] = get_default_kind(path)
    
    if 'area' not in front_matter:
        front_matter['area'] = get_default_area(path, content)

    if 'status' not in front_matter:
        front_matter['status'] = 'draft'
    
    if 'links' not in front_matter:
        front_matter['links'] = {}
    
    links = front_matter['links']
    if 'requirements' in front_matter:
        links['requirements'] = front_matter.pop('requirements')
    if 'tasks' in front_matter:
        links['tasks'] = front_matter.pop('tasks')
    if 'adrs' in front_matter:
        links['adrs'] = front_matter.pop('adrs')

    # Ensure requirements, tasks, adrs are arrays if they exist
    for k in ['requirements', 'tasks', 'adrs']:
        if k in links and not isinstance(links[k], list):
            links[k] = [links[k]]
        if k not in links:
            links[k] = []

    # Ensure doc_id
    # Always regenerate doc_id to ensure consistency and uniqueness across directories
    name_slug = slugify(path.stem)
    parent_parts = path.parent.relative_to(DOCS_DIR).parts
    if parent_parts:
        parent_slug = slugify("-".join(parent_parts))
        name_slug = f"{parent_slug}-{name_slug}"
    
    kind_slug = front_matter['kind'].upper()
    front_matter['doc_id'] = f"DOC-{kind_slug}-{name_slug}"

    # Remove extra fields
    allowed_fields = ['doc_id', 'kind', 'area', 'status', 'links']
    final_fm = {k: front_matter[k] for k in allowed_fields if k in front_matter}
    
    # Reorder fields for consistency
    ordered_fm = {}
    for k in allowed_fields:
        if k in final_fm:
            ordered_fm[k] = final_fm[k]
        else:
            # Should not happen given logic above
            pass

    new_content = "---\n" + yaml.dump(ordered_fm, sort_keys=False) + "---\n" + body
    
    with open(path, 'w', encoding='utf-8') as f:
        f.write(new_content)

if __name__ == '__main__':
    for md_file in DOCS_DIR.rglob('*.md'):
        print(f"Processing {md_file}")
        fix_file(md_file)

File: xtask/src/fix_front_matter.rs
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Documentation front matter repair and normalization.
//!
//! This is the Rust replacement for the former Python fixer. It preserves each
//! Markdown body while normalizing the metadata fields consumed by the docs
//! cross-reference tooling.

use anyhow::{Context, Result};
use serde_yaml::{Mapping, Value};
use std::path::Path;
use walkdir::WalkDir;

const DOCS_DIR: &str = "docs";

pub fn run_fix_front_matter() -> Result<()> {
    for entry in WalkDir::new(DOCS_DIR)
        .follow_links(false)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|entry| entry.path().is_file())
        .filter(|entry| entry.path().extension().and_then(|ext| ext.to_str()) == Some("md"))
    {
        let path = entry.into_path();
        println!("Processing {}", path.display());
        fix_file(&path)?;
    }

    Ok(())
}

fn fix_file(path: &Path) -> Result<()> {
    let content = std::fs::read_to_string(path)
        .with_context(|| format!("failed to read {}", path.display()))?;
    let (mut front_matter, body) = split_front_matter(&content)
        .with_context(|| format!("failed to parse front matter in {}", path.display()))?;

    normalize_legacy_field(&mut front_matter, "category", "kind", map_kind);
    normalize_legacy_field(&mut front_matter, "group", "area", map_area);

    if !contains_key(&front_matter, "kind") {
        set_value(
            &mut front_matter,
            "kind",
            Value::String(default_kind(path).to_string()),
        );
    }
    if !contains_key(&front_matter, "area") {
        set_value(
            &mut front_matter,
            "area",
            Value::String(default_area(&content).to_string()),
        );
    }
    if !contains_key(&front_matter, "status") {
        set_value(
            &mut front_matter,
            "status",
            Value::String("draft".to_string()),
        );
    }

    let mut links = take_mapping(&mut front_matter, "links").unwrap_or_default();
    move_legacy_link(&mut front_matter, &mut links, "requirements");
    move_legacy_link(&mut front_matter, &mut links, "tasks");
    move_legacy_link(&mut front_matter, &mut links, "adrs");
    for key in ["requirements", "tasks", "adrs"] {
        ensure_sequence(&mut links, key);
    }

    let kind = get_value(&front_matter, "kind")
        .and_then(Value::as_str)
        .unwrap_or("concept")
        .to_string();
    set_value(
        &mut front_matter,
        "doc_id",
        Value::String(format!("DOC-{}-{}", slugify(&kind), doc_slug(path)?)),
    );

    let mut ordered = Mapping::new();
    for key in ["doc_id", "kind", "area", "status"] {
        if let Some(value) = front_matter.remove(key) {
            ordered.insert(Value::String(key.to_string()), value);
        }
    }
    ordered.insert(Value::String("links".to_string()), Value::Mapping(links));

    let yaml = serde_yaml::to_string(&ordered).context("failed to serialize front matter")?;
    let new_content = format!("---\n{}---\n{}", yaml, body);
    std::fs::write(path, new_content).with_context(|| format!("failed to write {}", path.display()))
}

fn split_front_matter(content: &str) -> Result<(Mapping, &str)> {
    if let Some(stripped) = content.strip_prefix("---\n")
        && let Some(end) = stripped.find("\n---\n")
    {
        let yaml = &stripped[..end];
        let body = &stripped[end + "\n---\n".len()..];
        let map = serde_yaml::from_str::<Mapping>(yaml).context("invalid YAML front matter")?;
        return Ok((map, body));
    }

    Ok((Mapping::new(), content))
}

fn normalize_legacy_field(
    front_matter: &mut Mapping,
    old_key: &str,
    new_key: &str,
    mapper: fn(&str) -> &str,
) {
    if let Some(value) = remove_value(front_matter, old_key)
        && let Some(raw) = value.as_str()
    {
        front_matter.insert(
            Value::String(new_key.to_string()),
            Value::String(mapper(raw).to_string()),
        );
    }
}

fn move_legacy_link(front_matter: &mut Mapping, links: &mut Mapping, key: &str) {
    if let Some(value) = remove_value(front_matter, key) {
        links.insert(Value::String(key.to_string()), value);
    }
}

fn ensure_sequence(links: &mut Mapping, key: &str) {
    let key_value = Value::String(key.to_string());
    match links.remove(&key_value) {
        Some(Value::Sequence(values)) => {
            links.insert(key_value, Value::Sequence(values));
        }
        Some(Value::Null) | None => {
            links.insert(key_value, Value::Sequence(Vec::new()));
        }
        Some(value) => {
            links.insert(key_value, Value::Sequence(vec![value]));
        }
    }
}

fn take_mapping(front_matter: &mut Mapping, key: &str) -> Option<Mapping> {
    match remove_value(front_matter, key) {
        Some(Value::Mapping(mapping)) => Some(mapping),
        _ => None,
    }
}

fn remove_value(mapping: &mut Mapping, key: &str) -> Option<Value> {
    mapping.remove(Value::String(key.to_string()))
}

fn contains_key(mapping: &Mapping, key: &str) -> bool {
    mapping.contains_key(Value::String(key.to_string()))
}

fn get_value<'a>(mapping: &'a Mapping, key: &str) -> Option<&'a Value> {
    mapping.get(Value::String(key.to_string()))
}

fn set_value(mapping: &mut Mapping, key: &str, value: Value) {
    mapping.insert(Value::String(key.to_string()), value);
}

fn default_kind(path: &Path) -> &'static str {
    let parts: Vec<_> = path.components().collect();
    let has = |needle: &str| {
        parts
            .iter()
            .any(|part| part.as_os_str().to_string_lossy() == needle)
    };

    if has("explanation") {
        if has("adr") {
            return "adr";
        }
        return "explanation";
    }
    if has("how-to") {
        return "how-to";
    }
    if has("reference") {
        return "reference";
    }
    if has("tutorials") {
        return "tutorial";
    }
    if has("design") {
        return "design";
    }
    if has("dev") {
        return "explanation";
    }
    "concept"
}

fn default_area(content: &str) -> &'static str {
    let content = content.to_lowercase();
    for area in [
        "flight-core",
        "flight-virtual",
        "flight-hid",
        "flight-ipc",
        "flight-scheduler",
        "flight-ffb",
        "flight-panels",
        "infra",
        "ci",
        "simulation",
        "integration",
        "ksp",
        "profile",
    ] {
        if content.contains(area) {
            return area;
        }
    }
    "flight-core"
}

fn map_kind(value: &str) -> &str {
    match value {
        "explanation" => "explanation",
        "how-to" => "how-to",
        "reference" => "reference",
        "tutorial" => "tutorial",
        "design" => "design",
        "concept" => "concept",
        "requirements" => "requirements",
        "adr" => "adr",
        _ => "concept",
    }
}

fn map_area(value: &str) -> &str {
    match value {
        "flight-core" => "flight-core",
        "flight-virtual" => "flight-virtual",
        "flight-hid" => "flight-hid",
        "flight-ipc" => "flight-ipc",
        "flight-scheduler" => "flight-scheduler",
        "flight-ffb" => "flight-ffb",
        "flight-panels" => "flight-panels",
        "infra" | "infrastructure" => "infra",
        "ci" => "ci",
        "simulation" => "simulation",
        "integration" => "integration",
        "ksp" => "ksp",
        "profile" => "profile",
        _ => "flight-core",
    }
}

fn doc_slug(path: &Path) -> Result<String> {
    let mut slug_parts = Vec::new();
    if let Ok(relative_parent) = path
        .parent()
        .unwrap_or_else(|| Path::new(""))
        .strip_prefix(DOCS_DIR)
    {
        for part in relative_parent.components() {
            slug_parts.push(part.as_os_str().to_string_lossy().into_owned());
        }
    }
    let stem = path
        .file_stem()
        .and_then(|stem| stem.to_str())
        .map(str::to_string)
        .with_context(|| format!("failed to get file stem for {}", path.display()))?;
    slug_parts.push(stem);
    Ok(slugify(&slug_parts.join("-")))
}

fn slugify(value: &str) -> String {
    let mut slug = String::new();
    let mut last_was_dash = true;
    for ch in value.chars().flat_map(char::to_uppercase) {
        if ch.is_ascii_alphanumeric() {
            slug.push(ch);
            last_was_dash = false;
        } else if !last_was_dash {
            slug.push('-');
            last_was_dash = true;
        }
    }
    if slug.ends_with('-') {
        slug.pop();
    }
    slug
}
